hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- nilFile: a nilFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func readNilFile(ctx context.Context, file *nilFile, bucket string, target *probe.Target, client stiface.Client, logger *logger.Logger) (*nilFile, *probe.ProbeError) {
+	fileObject := client.Bucket(bucket).Object(file.name)
+	var reader stiface.Reader
+
+	err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APIGetFile, func() *probe.ProbeError {
+		var err error
+		reader, err = fileObject.NewReader(ctx)
+		if err != nil {
+			return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", err))
+
+		}
+		return probe.NewProbeError(metrics.Success, nil)
+	})
+	defer reader.Close()
+
+	if err.Status != metrics.Success {
+		return nil, err
+	}
+
+	logger.Infof("readNilFile(): nil file retrieved from target storage system.")
+
+	var readErr error
+	file.contents, readErr = ioutil.ReadAll(reader)
+	if readErr != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", readErr))
+	}
+
+	logger.Infof("readNilFile(): nil file read complete.")
+
+	return file, probe.NewProbeError(metrics.Success, nil)
+}
+
+// verifyNilFile verifies that the contents of the nil file match its checksum
+// and that the journal matches the existing in-memory proto.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- nilFile: a nilFile struct with the file name and contents recorded
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func verifyNilFile(ctx context.Context, file *nilFile, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	hash := sha1.New()
+	if _, err := io.Copy(hash, file.reader()); err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("verifyNilFile(): unable to compute checksum of file contents: %w", err))
+	}
+	file.hash = fmt.Sprintf("%x", hash.Sum(nil))
+
+	if !file.validateChecksum() {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): checksum of nil file did not match filename"))
+	}
+
+	journal := &sjpb.StateJournal{}
+	if err := proto.Unmarshal(file.contents, journal); err != nil {
+		return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("verifyNilFile(): unable to unmarshal nil file proto: %w", err))
+	}
+
+	logger.Infof("verifyNilFile(): unmarshal complete for nil file proto.")
+
+	if err := resolveIntent(ctx, journal, target, client, logger); err != nil {
+		return err
+	}
+
+	if diff := cmp.Diff(target.Journal, journal, protocmp.Transform()); diff != "" {
```

Comment:
This should be above `resolveIntent` - you probably don't want to resolve a corrupt intent.

Reply:
This compares the in-memory journal struct against the journal struct created from the contents of the file read from the target storage system. This aims to highlight any differences in state. If the intent is corrupt, this should show in the unmarshalling or during the resolution of the intent.